pytensor/loop/basic.py

```diff
@@ -0,0 +1,120 @@
+from typing import List, Tuple
+
+import numpy as np
+
+from pytensor import Variable, as_symbolic
+from pytensor.graph import FunctionGraph
+from pytensor.loop.op import Scan
+from pytensor.scan.utils import until
+from pytensor.tensor import as_tensor, empty_like
+
+
+def scan(
+    fn,
+    init_states=None,
+    sequences=None,
+    non_sequences=None,
+    n_steps=None,
+    go_backwards=False,
+) -> Tuple[List[Variable], List[Variable]]:
+    if sequences is None and n_steps is None:
+        raise ValueError("Must provide n_steps when scanning without sequences")
+
+    if init_states is None:
+        init_states = []
+    else:
+        if not isinstance(init_states, (tuple, list)):
+            init_states = [init_states]
+        init_states = [as_symbolic(i) for i in init_states]
+
+    if sequences is None:
+        sequences = []
+    else:
+        if not isinstance(sequences, (tuple, list)):
+            sequences = [sequences]
+        sequences = [as_tensor(s) for s in sequences]
+
+    if non_sequences is None:
+        non_sequences = []
+    else:
+        if not isinstance(non_sequences, (tuple, list)):
+            non_sequences = [non_sequences]
+        non_sequences = [as_symbolic(n) for n in non_sequences]
+
+    # Note: Old scan order is sequences + init + non_sequences
+    inner_sequences = [s[0] for s in sequences]
+    inner_inputs = [i.type() for i in init_states + inner_sequences + non_sequences]
```

Comment:
Hmm, I can't do this or else it won't be possible to reference the outer graph (e.g., to request gradients wrt non-sequences) as is don't by the jacobian function

Reply:
I don't get it yet. Can you give an example of what you'd want to do, but can't?

Reply:
This: https://pytensor.readthedocs.io/en/latest/tutorial/gradients.html#computing-the-jacobian

Note the box below where they say it's not possible to take the gradient with respect to the sequences, so it must be passed as a non-sequence. It's the same kind of problem, except in my first approach here I made it impossible to do so with the non-sequences as well. My approach was to replace all inputs by dummy inputs before calling the user provided function. That means a macro like `grad` cannot see the original graph and take the right gradient.

This however means I have to allow users to reference other variables (e.g., the inputs to that graph expression), which they are unlikely to pass as non-sequences (how would they know?)

I will change that. Maybe I can change it so that even the sequences are not dummy so that the approach that didn't use to work with the old scan would also work fine here. Once we have the "whole" graph we should be able to truncate it correctly (i.e, pass the minimum subset of `non-sequences` with the helper that @ferrine wrote for `replace`?

This is messy but I don't see another way around it. A large part of the old Scan initialization is indeed trying to deal with this:
https://github.com/pymc-devs/pytensor/blob/4da56ecf6c08cd5c5b960116b101ba6dde84a0db/pytensor/scan/basic.py#L890-L905

Reply:
Other discussion I found relevant for this:
https://groups.google.com/g/theano-users/c/fh9uYZnFj40/m/YpBQE5vtHQAJ

https://groups.google.com/g/theano-dev/c/h6G0QptsQOs/m/Hpx5B9EWO04J

Reply:
The whole point of the reverse loop was that we don't have to worry about that kind of thing. Shouldn't you just be able to use the gradients from the loop?
I guess that would only work if Scan is a subclass of OpFromGraph though, instead of an independent op that only knows about its Loop representation after a rewrite?

Reply:
This is not about the gradients of the Scan operations, it's about using a Scan to take gradients of other variables (e.g., to fill in the jacobian matrix). Anyway I pushed a new commit that fixes it